src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -3,22 +3,142 @@
 
 Description:
 This MicroService is meant to remove Rucio rules that are no longer needed in the Workload Management system, such as:
- * block-level rules created by WMAgent, against the origin RSE where data is getting produced
- * container and block-level rules created by MSTransferor, for input data that is no longer in the system
+ * block-level     rules created by WMAgent, against the origin RSE where data is getting produced
+ * container-level rules created by WMAgent, against the origin RSE where data is getting produced
+ * block-level     rules created by MSTransferor, for input data that is no longer in the system
+ * container-level rules created by MSTransferor, for input data that is no longer in the system
 In addition to that, this MicroService is now also responsible for the workflow archival, which is the final status that
 workflows remain.
 """
+# NOTE:
+#       The idea behind the construction of the code bellow is to make it work
+#       just like a Galton Board:
+#       https://en.wikipedia.org/wiki/Bean_machine
+# TODO: To Add the top level description of the logic into the module's Docstring
 
 # futures
 from __future__ import division, print_function
+import json
 
 # system modules
-from retry import retry
+from threading import current_thread
+from copy import deepcopy
+from pprint import pformat
 
 # WMCore modules
 from WMCore.MicroService.DataStructs.DefaultStructs import RULECLEANER_REPORT
 from WMCore.MicroService.Unified.MSCore import MSCore
+from WMCore.MicroService.Unified.Common import ckey, cert
+from WMCore.Services.pycurl_manager import RequestHandler
+from WMCore.ReqMgr.DataStructs import RequestStatus
 from Utils.EmailAlert import EmailAlert
+from Utils.Pipeline import Pipeline, Functor
+from Utils.Timers import timeFunction
+from WMCore.WMException import WMException
+
+
+class MSRuleCleanerException(WMException):
+    """
+    General Exception Class for MSRuleCleaner Module in WMCore MicroServices
+    """
+    def __init__(self, message):
+        self.myMessage = "MSRuleCleanerException: %s" % message
+        super(MSRuleCleanerException, self).__init__(self.myMessage)
+
+
+class MSRuleCleanerArchival(MSRuleCleanerException):
+    """
+    Archival Exception Class for MSRuleCleaner Module in WMCore MicroServices
+    """
+    def __init__(self, message):
+        self.myMessage = "MSRuleCleanerArchival: %s" % message
+        super(MSRuleCleanerArchival, self).__init__(self.myMessage)
+
+
+class MSRuleCleanerCleanup(MSRuleCleanerException):
```

Comment:
Yes indeed, I was planning to use it to signal a failed cleanup action per workflow from inside the `cleanRuciorules` function so that the counters for this workflows are not updated in a way similar to the `MSRulecleanerArchival` Exception.... but... as we can see the implementation did not follow the original design. I am deleting it now.